cmd/soroban-cli/src/commands/contract/fetch.rs

```diff
@@ -0,0 +1,235 @@
+            self.run_against_rpc_server().await
+        }
+    }
+
+    pub fn network(&self) -> Result<Network, Error> {
+        Ok(self.network.get(&self.locator)?)
+    }
+
+    pub async fn run_against_rpc_server(&self) -> Result<Vec<u8>, Error> {
+        let network = self.network()?;
+        tracing::trace!(?network);
+        let contract_id = self.contract_id()?;
+        let client = Client::new(&network.rpc_url)?;
+        // async closures are not yet stable
+        get_remote_wasm(&client, &contract_id).await
+    }
+
+    pub fn get_state(&self) -> Result<LedgerSnapshot, Error> {
+        Ok(self.ledger_file.read(&self.locator.config_dir()?)?)
+    }
+
+    pub fn run_in_sandbox(&self) -> Result<Vec<u8>, Error> {
+        let contract_id = self.contract_id()?;
+        // Initialize storage and host
+        let snap = Rc::new(self.get_state()?);
+        let mut storage = Storage::with_recording_footprint(snap);
+        Ok(get_contract_wasm_from_storage(&mut storage, contract_id)?)
+    }
+
+    fn contract_id(&self) -> Result<[u8; 32], Error> {
+        utils::contract_id_from_str(&self.contract_id)
+            .map_err(|e| Error::CannotParseContractId(self.contract_id.clone(), e))
+    }
+}
+
+async fn get_remote_wasm(client: &Client, contract_id: &[u8; 32]) -> Result<Vec<u8>, Error> {
+    // Get the contract from the network
+    let contract_key = LedgerKey::ContractData(LedgerKeyContractData {
+        contract_id: xdr::Hash(*contract_id),
+        key: ScVal::LedgerKeyContractExecutable,
+    });
+    let contract_ref = client.get_ledger_entries(Vec::from([contract_key])).await?;
+    if contract_ref.entries.is_empty() {
+        return Err(Error::MissingResult);
+    }
+    let contract_ref_entry = &contract_ref.entries[0];
+    Ok(
+        match LedgerEntryData::from_xdr_base64(&contract_ref_entry.xdr)? {
+            LedgerEntryData::ContractData(ContractDataEntry {
+                val: ScVal::ContractExecutable(ScContractExecutable::WasmRef(hash)),
+                ..
+            }) => {
+                let code_key = LedgerKey::ContractCode(LedgerKeyContractCode { hash });
+                let contract_data = client.get_ledger_entries(Vec::from([code_key])).await?;
+                if contract_data.entries.is_empty() {
+                    return Err(Error::MissingResult);
+                }
+                let contract_data_entry = &contract_data.entries[0];
+                match LedgerEntryData::from_xdr_base64(&contract_data_entry.xdr)? {
+                    LedgerEntryData::ContractCode(ContractCodeEntry { code, .. }) => code.into(),
+                    scval => return Err(Error::UnexpectedContractCodeDataType(scval)),
+                }
+            }
+            scval => return Err(Error::UnexpectedContractCodeDataType(scval)),
+        },
+    )
+}
+
+pub fn get_contract_wasm_from_storage(
+    storage: &mut Storage,
+    contract_id: [u8; 32],
+) -> Result<Vec<u8>, FromWasmError> {
```

Comment:
I'm assuming this (and `get_remote_wasm` above) are public so they can be re-used by other subcommands (like `inspect`)? If so, we should put them somewhere else. Having subcommands depending on pieces from other subcommands is going to turn this into a mess quickly.

It would be nice if we had a way to put this method onto the `Storage`, but barring that it could go into the `utils` package, where we already have a `get_contract_spec_from_storage` function.

Reply:
I'll move it to util.rs for now, but I agree that the presence of a util suggests we should move the needed functions upstream.